rest-client/client-vertx-impl/src/main/java/io/apicurio/registry/rest/client/VertxHttpClient.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2021 Red Hat
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.apicurio.registry.rest.client;
+
+import io.apicurio.registry.auth.Auth;
+import io.apicurio.registry.rest.client.config.ClientConfig;
+import io.apicurio.registry.rest.client.exception.RestClientException;
+import io.apicurio.registry.rest.client.impl.ErrorHandler;
+import io.apicurio.registry.rest.client.request.Request;
+import io.apicurio.registry.rest.client.response.ResponseHandler;
+import io.apicurio.registry.utils.IoUtil;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.http.RequestOptions;
+import io.vertx.ext.web.client.HttpRequest;
+import io.vertx.ext.web.client.WebClient;
+import org.apache.http.NameValuePair;
+import org.apache.http.client.utils.URIBuilder;
+import org.keycloak.authorization.client.util.HttpResponseException;
+
+import java.io.UncheckedIOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+public class VertxHttpClient implements RegistryHttpClient {
+
+    private final WebClient webClient;
+    private final Auth auth;
+    private final String basePath;
+    private final Vertx vertx;
+
+    private static final Map<String, String> DEFAULT_HEADERS = new HashMap<>();
+    private static final ThreadLocal<Map<String, String>> requestHeaders = ThreadLocal.withInitial(Collections::emptyMap);
```

Comment:
This sounds like a bad idea. On an event loop, executions of independent actions will interleave on the same thread.